File: fc/internal/timeutil/timeutil_test.go
package timeutil

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
	"time"
)

type dayBoundCase struct {
	Name     string `json:"name"`
	TZ       string `json:"tz"`
	Year     int    `json:"year"`
	Month    int    `json:"month"`
	Day      int    `json:"day"`
	StartUTC string `json:"startUtc"`
	EndUTC   string `json:"endUtc"`
}

func loadDayBoundCases(t *testing.T) []dayBoundCase {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	root := filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", ".."))
	b, err := os.ReadFile(filepath.Join(root, "testdata", "zoned-day-bounds-cases.json"))
	if err != nil {
		t.Fatalf("read zoned-day-bounds-cases: %v", err)
	}
	var payload struct {
		Cases []dayBoundCase `json:"cases"`
	}
	if err := json.Unmarshal(b, &payload); err != nil {
		t.Fatalf("parse zoned-day-bounds-cases: %v", err)
	}
	return payload.Cases
}

func TestExpandCompactOffset(t *testing.T) {
	cases := map[string]string{
		"2026-07-30T08:00:00+0800":     "2026-07-30T08:00:00+08:00",
		"2026-07-30T08:00:00.123-0530": "2026-07-30T08:00:00.123-05:30",
		"2026-07-30T08:00:00+08:00":    "2026-07-30T08:00:00+08:00",
		"2026-07-30T00:00:00.000Z":     "2026-07-30T00:00:00.000Z",
	}
	for in, want := range cases {
		if got := ExpandCompactOffset(in); got != want {
			t.Fatalf("%q: got %q want %q", in, got, want)
		}
	}
	got, err := ParseRFC3339Flexible("2026-07-30T08:00:00+0800")
	if err != nil {
		t.Fatal(err)
	}
	want, _ := time.Parse(time.RFC3339, "2026-07-30T08:00:00+08:00")
	if !got.Equal(want) {
		t.Fatalf("instant: got %v want %v", got, want)
	}
	for _, bad := range []string{
		"2026-07-30T08:00:00z",
		"2026-07-30 08:00:00Z",
		"2026-7-30T08:00:00Z",
		"2026-07-30T8:00:00Z",
	} {
		if _, err := ParseRFC3339Flexible(bad); err == nil {
			t.Fatalf("want reject %q", bad)
		}
	}
}

func TestEmbeddedTzdataWithoutSystemZoneinfo(t *testing.T) {
	// ZONEINFO 指向不存在路径时，须仍能靠 //go:embed 的 time/tzdata 加载
	t.Setenv("ZONEINFO", filepath.Join(t.TempDir(), "no-such-zoneinfo"))
	if !IsValidTimeZone("Asia/Shanghai") {
		t.Fatal("Asia/Shanghai should load from embedded time/tzdata")
	}
	now := time.Date(2026, 7, 30, 16, 30, 0, 0, time.UTC)
	start, end, err := GetZonedDayBounds(now, "Asia/Shanghai")
	if err != nil {
		t.Fatal(err)
	}
	if start.UTC().Format(time.RFC3339) != "2026-07-30T16:00:00Z" {
		t.Fatalf("shanghai start: %s", start.UTC())
	}
	if end.UTC().Format(time.RFC3339) != "2026-07-31T16:00:00Z" {
		t.Fatalf("shanghai end: %s", end.UTC())
	}
}

func TestIsValidTimeZone(t *testing.T) {
	if !IsValidTimeZone("Asia/Shanghai") || !IsValidTimeZone("UTC") {
		t.Fatal("expected valid zones")
	}
	if IsValidTimeZone("Not/AZone") || IsValidTimeZone("") {
		t.Fatal("expected invalid zones")
	}
	// Go LoadLocation 会收下、Intl 拒绝的非 IANA 名 → API 层须拒绝，避免双端分叉
	for _, tz := range []string{"Factory", "localtime", "posixrules"} {
		if IsValidTimeZone(tz) {
			t.Fatalf("want reject Go-only zone %q", tz)
		}
	}
}

func TestCalendarDayBoundsSharedFixtures(t *testing.T) {
	for _, c := range loadDayBoundCases(t) {
		start, end, err := CalendarDayBounds(c.Year, c.Month, c.Day, c.TZ)
		if err != nil {
			t.Fatalf("%s: %v", c.Name, err)
		}
		gotStart := start.UTC().Format("2006-01-02T15:04:05.000Z07:00")
		gotEnd := end.UTC().Format("2006-01-02T15:04:05.000Z07:00")
		if gotStart != c.StartUTC {
			t.Fatalf("%s start: got %s want %s", c.Name, gotStart, c.StartUTC)
		}
		if gotEnd != c.EndUTC {
			t.Fatalf("%s end: got %s want %s", c.Name, gotEnd, c.EndUTC)
		}
	}
}

func TestGetZonedDayBounds(t *testing.T) {
	// 2026-07-30 16:30 UTC = 上海 7/31 00:30 → 日历日 7/31
	now := time.Date(2026, 7, 30, 16, 30, 0, 0, time.UTC)

	start, end, err := GetZonedDayBounds(now, "Asia/Shanghai")
	if err != nil {
		t.Fatal(err)
	}
	if start.UTC().Format(time.RFC3339) != "2026-07-30T16:00:00Z" {
		t.Fatalf("shanghai start: %s", start.UTC())
	}
	if end.UTC().Format(time.RFC3339) != "2026-07-31T16:00:00Z" {
		t.Fatalf("shanghai end: %s", end.UTC())
	}

	start, end, err = GetZonedDayBounds(now, "UTC")
	if err != nil {
		t.Fatal(err)
	}
	if start.UTC().Format(time.RFC3339) != "2026-07-30T00:00:00Z" {
		t.Fatalf("utc start: %s", start.UTC())
	}
	if end.UTC().Format(time.RFC3339) != "2026-07-31T00:00:00Z" {
		t.Fatalf("utc end: %s", end.UTC())
	}
}

File: src/lib/timeutil.test.ts
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import {
  calendarDayBounds,
  expandCompactOffset,
  getZonedDayBounds,
  isValidTimeZone,
  parseRFC3339Flexible,
} from './timeutil'

type DayBoundCase = {
  name: string
  tz: string
  year: number
  month: number
  day: number
  startUtc: string
  endUtc: string
}

const dayBoundCases = (
  JSON.parse(
    readFileSync(
      path.join(
        path.dirname(fileURLToPath(import.meta.url)),
        '../../testdata/zoned-day-bounds-cases.json',
      ),
      'utf8',
    ),
  ) as { cases: DayBoundCase[] }
).cases

describe('time helpers', () => {
  it('rejects invalid IANA time zones', () => {
    expect(isValidTimeZone('Asia/Shanghai')).toBe(true)
    expect(isValidTimeZone('UTC')).toBe(true)
    expect(isValidTimeZone('Not/AZone')).toBe(false)
    expect(isValidTimeZone('')).toBe(false)
    // Go LoadLocation 会收下的非 IANA 名；Intl 与 API 均须拒绝
    expect(isValidTimeZone('Factory')).toBe(false)
    expect(isValidTimeZone('localtime')).toBe(false)
    expect(isValidTimeZone('posixrules')).toBe(false)
  })

  it('expandCompactOffset / parseRFC3339Flexible match Go', () => {
    expect(expandCompactOffset('2026-07-30T08:00:00+0800')).toBe(
      '2026-07-30T08:00:00+08:00',
    )
    expect(expandCompactOffset('2026-07-30T00:00:00.000Z')).toBe(
      '2026-07-30T00:00:00.000Z',
    )
    const ok = parseRFC3339Flexible('2026-07-30T08:00:00+0800')
    expect(ok?.toISOString()).toBe('2026-07-30T00:00:00.000Z')
    expect(parseRFC3339Flexible('2026-07-30T08:00:00z')).toBeNull()
    expect(parseRFC3339Flexible('2026-07-30 08:00:00Z')).toBeNull()
    expect(parseRFC3339Flexible('2026-7-30T08:00:00Z')).toBeNull()
    expect(parseRFC3339Flexible('2026-07-30T8:00:00Z')).toBeNull()
  })

  it('matches shared zoned-day-bounds fixtures (incl. DST)', () => {
    for (const c of dayBoundCases) {
      const { start, end } = calendarDayBounds(c.year, c.month, c.day, c.tz)
      expect(start.toISOString(), c.name).toBe(c.startUtc)
      expect(end.toISOString(), c.name).toBe(c.endUtc)
    }
  })

  it('getZonedDayBounds uses the calendar day of now in zone', () => {
    // 2026-07-30 16:30 UTC = 2026-07-31 00:30 Asia/Shanghai → 上海日历日 7/31
    const now = new Date('2026-07-30T16:30:00.000Z')
    const shanghai = getZonedDayBounds(now, 'Asia/Shanghai')
    expect(shanghai.start.toISOString()).toBe('2026-07-30T16:00:00.000Z')
    expect(shanghai.end.toISOString()).toBe('2026-07-31T16:00:00.000Z')
  })
})
